# Configuration
pepfile: "config/config.yaml"

# Variables
import itertools
CONTRASTS = list(config["contrasts"].keys())
GROUPS = list(config["groups"].keys())
SAMPLES = list(config["samples"].keys())
RUNS = config["samples"].values()
RUNS = list(itertools.chain.from_iterable(RUNS))

# Wildcards
wildcard_constraints:
    contrasts = "|".join(CONTRASTS),
    groups = "|".join(GROUPS),
    sample = "|".join(SAMPLES),
    run = "|".join(RUNS)

# Targets
rule default:
    input:
        expand("results/bwa/{sample}/{subsample}.bam")

# Rules
include: "workflow/rules/bwa.smk"
include: "workflow/rules/cutadapt.smk"
include: "workflow/rules/fastqc.smk"
include: "workflow/rules/genomepy.smk"
include: "workflow/rules/samtools.smk"